pkg/exchange/coinbase/stream.go

```diff
@@ -25,10 +37,155 @@ type Stream struct {
+		passphrase:     passphrase,
+		secretKey:      secretKey,
 	}
 	s.SetParser(s.parseMessage)
+	s.SetDispatcher(s.dispatchEvent)
+	s.SetEndpointCreator(createEndpoint)
+
+	// public handlers
+	s.OnConnect(s.handleConnect)
+
 	return &s
 }
+
+func (s *Stream) dispatchEvent(e interface{}) {
+	switch e := e.(type) {
+	case *StatusMessage:
+		s.EmitStatusMessage(e)
+	case *AuctionMessage:
+		s.EmitAuctionMessage(e)
+	case *RfqMessage:
+		s.EmitRfqMessage(e)
+	case *TickerMessage:
+		s.EmitTickerMessage(e)
+	case *ReceivedLimitOrderMessage:
+		s.EmitReceivedLimitOrderMessage(e)
+	case *ReceivedMarketOrderMessage:
+		s.EmitReceivedMarketOrderMessage(e)
+	case *OpenMessage:
+		s.EmitOpenMessage(e)
+	case *DoneMessage:
+		s.EmitDoneMessage(e)
+	case *MatchMessage:
+		s.EmitMatchMessage(e)
+	case *AuthMakerMatchMessage:
+		s.EmitAuthMakerMatchMessage(e)
+	case *AuthTakerMatchMessage:
+		s.EmitAuthTakerMatchMessage(e)
+	case *StpChangeMessage:
+		s.EmitStpChangeMessage(e)
+	case *ModifyOrderChangeMessage:
+		s.EmitModifyOrderChangeMessage(e)
+	case *ActiveMessage:
+		s.EmitActiveMessage(e)
+	default:
+		log.Warnf("skip dispatching msg due to unknown message type: %T", e)
+	}
+}
+
+func createEndpoint(ctx context.Context) (string, error) {
+	return wsFeedURL, nil
+}
+
+type channelType struct {
+	Name       string   `json:"name"`
+	ProductIDs []string `json:"product_ids,omitempty"`
+}
+
+type websocketCommand struct {
+	Type       string        `json:"type"`
+	Channels   []channelType `json:"channels"`
+	Signature  *string       `json:"signature,omitempty"`
+	Key        *string       `json:"key,omitempty"`
+	Passphrase *string       `json:"passphrase,omitempty"`
+	Timestamp  *string       `json:"timestamp,omitempty"`
+}
+
+func (s *Stream) handleConnect() {
+	// subscribe to channels
+	if len(s.Subscriptions) == 0 {
+		return
+	}
+
+	subProductsMap := make(map[string][]string)
+	for _, sub := range s.Subscriptions {
+		strChannel := string(sub.Channel)
+		// "rfq_matches" allow empty symbol
+		if sub.Channel != "rfq_matches" && len(sub.Symbol) == 0 {
+			continue
+		}
+		products := append(subProductsMap[strChannel], sub.Symbol)
+		subProductsMap[strChannel] = products
+	}
+	subCmds := []websocketCommand{}
+	signature, ts := s.generateSignature()
+	for channel, productIDs := range subProductsMap {
+		var subType string
+		switch channel {
+		case "rfq_matches":
+			subType = "subscriptions"
+		default:
+			subType = "subscribe"
+		}
+		subCmd := websocketCommand{
+			Type: subType,
+			Channels: []channelType{
+				{
+					Name:       channel,
+					ProductIDs: productIDs,
+				},
+			},
+		}
+		if s.authEnabled() {
+			subCmd.Signature = &signature
+			subCmd.Key = &s.apiKey
+			subCmd.Passphrase = &s.passphrase
+			subCmd.Timestamp = &ts
+		}
+		subCmds = append(subCmds, subCmd)
+	}
+	for _, subCmd := range subCmds {
+		err := s.Conn.WriteJSON(subCmd)
+		if err != nil {
+			log.WithError(err).Errorf("subscription error: %v", subCmd)
+		}
+	}
```

Comment:
Do we meet its requirements?

https://docs.cdp.coinbase.com/exchange/docs/websocket-rate-limits#subscription-limits

Reply:
The subscription may be rejected if we exceed the limits.
I think the program should keep on running so I simply add a warning log here.